### PR TITLE
MINOR: Tolerate limited data loss for upgrade tests with old message format

### PR DESCRIPTION
**Pull request by rajinisivaram**
We see transient test failures in system tests for upgrade scenarios when older message formats prior to KIP-101 are used. Since we need to continue to test these versions, tolerate a small amount of data loss in those tests. We will still be testing that upgrade works and most messages are successfully delivered.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: tests/kafkatest/utils/util.py [hunk 1]
```diff
@@ -155,7 +155,13 @@ def validate_delivery(acked, consumed, idempotence_enabled=False, check_lost_dat
         if check_lost_data:
             to_validate = list(missing)[0:1000 if len(missing) > 1000 else len(missing)]
             data_lost = check_lost_data(to_validate)
-            msg = annotate_data_lost(data_lost, msg, len(to_validate))
+
+            # With older versions of message format before KIP-101, data loss could occur due to truncation.
+            # These records won't be in the data logs. Tolerate limited data loss for this case.
+            if len(data_lost) == len(to_validate) and len(missing) <= data_loss_tolerance:
```
Comment (ijuma) [inline tests/kafkatest/utils/util.py]: What does the first condition (`len(data_lost) == len(to_validate)`) achieve?

Comment (rajinisivaram) [reply tests/kafkatest/utils/util.py]: @ijuma Thanks for the review. The specific case I was trying to fix was when the log was truncated and hence acked records not received by the consumer are not in the log. The first condition checks that. We will fail the test if consumer doesn't receive records that are present in the log. Does that make sense?

Comment (ijuma) [reply tests/kafkatest/utils/util.py]: Would we want to check all of `missing` and not just the `to_validate` subset in this case? Also, it's a bit confusing to set `success` to `True`, then to `False` and then set it back to `True` again. Can we refactor the logic a bit?

Comment (rajinisivaram) [reply tests/kafkatest/utils/util.py]: @ijuma Updated the code to make it more obvious that we are checking all of missing and not a subset.

File: tests/kafkatest/utils/util.py [hunk 2]
```diff
@@ -149,13 +150,25 @@ def validate_delivery(acked, consumed, idempotence_enabled=False, check_lost_dat
...
+
+            to_validate = list(missing)[0:min(len(missing), max_validate_count)]
             data_lost = check_lost_data(to_validate)
-            msg = annotate_data_lost(data_lost, msg, len(to_validate))
+
+            # With older versions of message format before KIP-101, data loss could occur due to truncation.
+            # These records won't be in the data logs. Tolerate limited data loss for this case.
+            if len(missing) < max_truncate_count and len(data_lost) == len(missing):
+               msg += "The %s missing messages were not present in Kafka's data files. This suggests data loss " \
+                   "due to truncation. The messages lost: %s\n" % (len(data_lost), str(data_lost))
```
Comment (ijuma) [inline tests/kafkatest/utils/util.py]: Can we clarify in the message that this is OK (like you did in the comment above)?

**Review by ijuma — APPROVED** [review]
LGTM, just a minor comment with regards to one of the text messages.

**Comment by rajinisivaram** [comment]
@ijuma Thanks for the review, merging to trunk.

**Pull request merged**